tools/filetools.go

```diff
@@ -121,3 +126,250 @@ func VerifyFileHash(oid, path string) error {
+			return true
+		}
+
+	}
+	// Also support matching a parent directory without a wildcard
+	if strings.HasPrefix(name, pattern+string(filepath.Separator)) {
+		return true
+	}
+
+	return false
+
+}
+
+// Returned from FastWalk with parent directory context
+// This is needed because FastWalk can provide paths out of order so the
+// parent dir cannot be implied
+type FastWalkInfo struct {
+	ParentDir string
+	Info      os.FileInfo
+}
+
+// FastWalk is a more optimal implementation of filepath.Walk
+// It differs in the following ways:
+//  * Provides a channel of information instead of using a callback func
+//  * Uses goroutines to parallelise large dirs and descent into subdirs
+//  * Does not provide sorted output; parents will always be before children but
+//    there are no other guarantees. Use parentDir in the FastWalkInfo struct to
+//    determine absolute path rather than tracking it yourself like filepath.Walk
+//  * Supports include / exclude filters
+// Both dir and include/exclude paths can be relative or absolute, but they must
+// all be of the same type. includePaths/excludePaths can be nil.
+func FastWalk(dir string, includePaths, excludePaths []string) (<-chan FastWalkInfo, <-chan error) {
+	return FastWalkWithExcludeFiles(dir, "", includePaths, excludePaths)
+}
+
+// FastWalkWithExcludeFiles is like FastWalk but with the additional option to
+// load any file named excludeFilename in any directory, and add its contents
+// to the excludePaths list for that directory and children.
+func FastWalkWithExcludeFiles(dir, excludeFilename string,
+	includePaths, excludePaths []string) (<-chan FastWalkInfo, <-chan error) {
+	fiChan := make(chan FastWalkInfo, 256)
+	errChan := make(chan error, 10)
+
+	go fastWalkFromRoot(dir, excludeFilename, includePaths, excludePaths, fiChan, errChan)
+
+	return fiChan, errChan
+}
+
+// FastWalkGitRepo behaves like FastWalkWithExcludeFiles, preconfigured to ignore
+// the git repo itself (.git) and to load exclude patterns from .gitignore
+func FastWalkGitRepo(dir string) (<-chan FastWalkInfo, <-chan error) {
+	// Ignore all git metadata including subrepos
+	excludePaths := []string{".git", filepath.Join("**", ".git")}
+	return FastWalkWithExcludeFiles(dir, ".gitignore", nil, excludePaths)
+}
+
+func fastWalkFromRoot(dir string, excludeFilename string,
+	includePaths, excludePaths []string, fiChan chan<- FastWalkInfo, errChan chan<- error) {
+
+	dirFi, err := os.Stat(dir)
+	if err != nil {
+		errChan <- err
+		return
+	}
+
+	// This waitgroup will be incremented for each nested goroutine
+	var waitg sync.WaitGroup
+
+	fastWalkItem(filepath.Dir(dir), dirFi, excludeFilename, includePaths, excludePaths, fiChan, errChan, &waitg)
+
+	waitg.Wait()
+	close(fiChan)
+	close(errChan)
+
+}
+
+// Main recursive implementation of fast walk
+// Increment waitg.Add(1) for each new goroutine launched internally
+func fastWalkItem(parentDir string, itemFi os.FileInfo, excludeFilename string,
```

Comment:
> Is an item a directory? A file? Something else?

It's both, that's why I used `Item` and not `File` or `Dir`. Guess it could be `fastWalkFileOrDir`